src/codegen/llvm/llvm_utils.cpp

```diff
@@ -61,6 +63,71 @@ static void run_optimisation_passes(llvm::Module& module,
 /*                             Optimisation utils                                       */
 /****************************************************************************************/
 
+void initialise_nvptx_passes() {
+    // Register targets.
+    LLVMInitializeNVPTXTarget();
+    LLVMInitializeNVPTXTargetMC();
+    LLVMInitializeNVPTXTargetInfo();
+    LLVMInitializeNVPTXAsmPrinter();
+
+    // Initialize passes.
+    initialise_optimisation_passes();
+}
+
+void optimise_module_for_nvptx(llvm::Module& module, int opt_level, std::string& target_asm) {
+    // CUDA target machine we generating code for.
+    std::unique_ptr<llvm::TargetMachine> tm;
+
+    // Hardcode target infromation for now. Change if necessary.
+    llvm::Triple triple("nvptx64-nvidia-cuda");
+    std::string subtarget = "sm_20";
+    std::string features = "+ptx60";
```

Comment:
IIUC this selects the PTX ISA version. I guess that we can use something more recent like `ptx70`?
Not sure whether this makes any difference in the generated code and performance though. Maybe it's something to check.
Also from what I've seen selecting an invalid version gives the earliest PTX supported version.